barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/goblin_ultra_circuit_builder.hpp

```diff
@@ -0,0 +1,113 @@
+#pragma once
+#include "barretenberg/plonk/proof_system/constants.hpp"
+#include "barretenberg/plonk/proof_system/types/polynomial_manifest.hpp"
+#include "barretenberg/plonk/proof_system/types/prover_settings.hpp"
+#include "barretenberg/polynomials/polynomial.hpp"
+#include "barretenberg/proof_system/arithmetization/arithmetization.hpp"
+#include "barretenberg/proof_system/op_queue/ecc_op_queue.hpp"
+#include "barretenberg/proof_system/plookup_tables/plookup_tables.hpp"
+#include "barretenberg/proof_system/plookup_tables/types.hpp"
+#include "barretenberg/proof_system/types/merkle_hash_type.hpp"
+#include "barretenberg/proof_system/types/pedersen_commitment_type.hpp"
+#include "ultra_circuit_builder.hpp"
+#include <optional>
+
+namespace proof_system {
+
+using namespace barretenberg;
+
+template <typename FF> class GoblinUltraCircuitBuilder_ : public UltraCircuitBuilder_<FF> {
+  public:
+    static constexpr std::string_view NAME_STRING = "GoblinUltraArithmetization";
+    static constexpr CircuitType CIRCUIT_TYPE = CircuitType::ULTRA;
+    static constexpr size_t DEFAULT_NON_NATIVE_FIELD_LIMB_BITS =
+        UltraCircuitBuilder_<FF>::DEFAULT_NON_NATIVE_FIELD_LIMB_BITS;
+
+    size_t num_ecc_op_gates = 0; // number of ecc op "gates" (rows); these are placed at the start of the circuit
+
+    // Stores record of ecc operations and performs corresponding native operations internally
+    std::shared_ptr<ECCOpQueue> op_queue;
+
+    // Indices for constant variables corresponding to ECCOpQueue op codes
+    uint32_t null_op_idx;
+    uint32_t add_accum_op_idx;
+    uint32_t mul_accum_op_idx;
+    uint32_t equality_op_idx;
+
+    using WireVector = std::vector<uint32_t, ContainerSlabAllocator<uint32_t>>;
+
+    // Wires storing ecc op queue data; values are indices into the variables array
+    std::array<WireVector, arithmetization::Ultra<FF>::NUM_WIRES> ecc_op_wires;
+
+    WireVector& ecc_op_wire_1 = std::get<0>(ecc_op_wires);
+    WireVector& ecc_op_wire_2 = std::get<1>(ecc_op_wires);
+    WireVector& ecc_op_wire_3 = std::get<2>(ecc_op_wires);
+    WireVector& ecc_op_wire_4 = std::get<3>(ecc_op_wires);
+
+    ecc_op_tuple queue_ecc_add_accum(const g1::affine_element& point);
+    ecc_op_tuple queue_ecc_mul_accum(const g1::affine_element& point, const FF& scalar);
+    ecc_op_tuple queue_ecc_eq();
+
+  private:
+    void populate_ecc_op_wires(const ecc_op_tuple& in);
```

Comment:
These functions are not documented

Reply:
The documentation for these is in the source (which I generally prefer) but I added a brief comment here: `// Functions for adding ECC op queue "gates"`